[GSoC] Modify point_on_surface methods in wrapping geometries

RushabhMehta2005 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
This PR is a result of the [discussion](https://github.com/sympy/sympy/pull/28135/files#r2156208314) between @moorepants and I upon the use of ``point_on_surface`` method
in both ``WrappingSphere`` and ``WrappingCylinder`` classes.

#### Brief description of what is fixed or changed
I have modified it to return an unevaluated symbolic equality rather than a forced boolean.  
Additionally I have removed the calls to it in all other methods i.e ``geodesic_length`` and ``geodesic_end_vectors``.

#### Other comments
This PR is a part of my GSoC 2025 work on the project "Implement Wrapping Geometries and Pathways for Musculoskeletal Modeling".

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

RushabhMehta2005 committed:
Deprecate point_on_surface methods for both WrappingSphere and WrappingCylinder

RushabhMehta2005 committed:
Fix trailing whitespaces

RushabhMehta2005 commented:
I know I am supposed to add a release notes entry manually after this gets merged, thus I left it as NO ENTRY for now.

I am quite confused on how to add tests for the deprecated method, as virtually every test will need the warning.

eh111eh commented:
> I know I am supposed to add a release notes entry manually after this gets merged, thus I left it as NO ENTRY for now.
> 
> I am quite confused on how to add tests for the deprecated method, as we are also testing for ValueError in some of them.

I think we need to test
1. Warning raise to ensure the method raises the appropriate warning (DeprecationWarning) which I think can be tested like
```python
@pytest.mark.parametrize("position", [Symbol('r') * Point('pO')])
    def test_deprecation_warning(self, position):
        r = Symbol('r', positive=True)
        pO = Point('pO')
        sphere = WrappingSphere(r, pO)

        with pytest.warns(DeprecationWarning, match="Checking if a point lies on sphere's surface by calling point_on_surface() is deprecated"):
            sphere.point_on_surface(position)
```
2. Correct behavior to
    i. Verify that the new wrapping geometries work correctly mathematically, physically, and geometrically, without relying on point_on_surface() by
```python
@pytest.mark.parametrize(
        'position_1, position_2, expected',
        [
            (r*N.x, r*N.x, S.Zero),
            ...
        ]
    )
    def test_geodesic_length_without_point_on_surface(self, position_1, position_2, expected):
        r = Symbol('r', positive=True)
        pO = Point('pO')
        sphere = WrappingSphere(r, pO)

        p1 = Point('p1')
        p1.set_pos(pO, position_1)
        p2 = Point('p2')
        p2.set_pos(pO, position_2)

        assert sphere.geodesic_length(p1, p2) == expected

@pytest.mark.parametrize(...)
    def test_geodesic_end_vectors_without_point_on_surface(self, position_1, position_2, vector_1, vector_2):
        r = Symbol('r', positive=True)
        pO = Point('pO')
        sphere = WrappingSphere(r, pO)

        p1 = Point('p1')
        p1.set_pos(pO, position_1)
        p2 = Point('p2')
        p2.set_pos(pO, position_2)

        assert sphere.geodesic_end_vectors(p1, p2) == (vector_1, vector_2)

```

    ii. Check that the expected exception (e.g., ValueError) is raised when points are invalid for geodesic calculations by

```python
@pytest.mark.parametrize('position', [S.Zero, 2*r*N.x])
    def test_geodesic_length_invalid_points(self, position):
        r = Symbol('r', positive=True)
        pO = Point('pO')
        sphere = WrappingSphere(r, pO)

        p1 = Point('p1')
        p1.set_pos(pO, position)
        p2 = Point('p2')
        p2.set_pos(pO, position)

        with pytest.raises(ValueError):
            sphere.geodesic_length(p1, p2)
```

eh111eh commented:
Can we also clean up or clarify the docstrings for geodesic_length() and geodesic_end_vectors()? They still reference point_on_surface() or mention points being on the surface, like:

"If the point is not on the surface, a ValueError is raised…"
which could be rephrased like
"This method assumes that the input points lie on the surface. It is the user's responsibility to ensure this…"

RushabhMehta2005 commented:
Thank you so much for your review. I'll incorporate these changes immediately.

RushabhMehta2005 committed:
Add neccesary checks back, fix docstrings, add deprecation warnings in pytest

RushabhMehta2005 commented:
> Also, can we delete `test_point_is_on_surface` function as well? I think the rest of the tests in both classes are not affected by or dependent on point_on_surface().

Oh so we will not test deprecated functionality anymore.

RushabhMehta2005 commented:
@moorepants Could you please review this ? 
We want to make sure that we have followed all the neccessary steps to deprecate a method. Except adding this deprecation to release notes manually.

oscarbenjamin commented on doc/src/explanation/active-deprecations.md:
```diff
@@ -76,7 +76,21 @@ SymPy deprecation warnings.
+(deprecated-point-on-surface-check-sphere)=
+### Deprecated point_on_surface method in WrappingSphere
+
+This method has been deprecated as it is deemed no longer necessary
+to check whether input points to ``WrappingSphere.geodesic_length`` and
+``WrappingSphere.geodesic_end_vectors`` lie on the surface of the sphere
+or not. These methods now rely on the user to supply valid points.
```
The fact that the method is not useful is not necessarily a reason for deprecating it. The method could be useful for some people sometimes even if it is not going to be used internally in sympy.

RushabhMehta2005 replied:
Other than the reason of deprecation, is it deprecated properly?

RushabhMehta2005 replied:
Alright, the calls to this method are already removed, maybe that is enough for the time being? 
@eh111eh I would love to hear your thoughts on this approach.

eh111eh replied:
Alternatively, we can keep using the method but modify it to return the symbolic condition itself rather than a boolean. That way, users can decide how to interpret the result: Whether to use their points based on the condition or apply simplification as they see fit.

RushabhMehta2005 committed:
Remove deprecation from point_on_surface methods.
Make point_on_surface methods return an unevaluated symbolic equality instead of boolean.

RushabhMehta2005 commented:
I think this one is also ready to be reviewed and merged if it looks good to you @eh111eh.

eh111eh commented on sympy/physics/mechanics/tests/test_wrapping_geometry.py:
```diff
@@ -23,7 +23,6 @@
 )
 from sympy.simplify.simplify import simplify
 
-
```
What is this indentation?

eh111eh reviewed (commented):
Looks good to me, except for the indentation. This way, we can keep using the method but modify it to return the symbolic condition itself rather than a boolean, so users can decide how/whether to use the result.

What do you think? @oscarbenjamin @moorepants

RushabhMehta2005 committed:
Add newline
